### PR TITLE
Closes #2668: Add click to copy functionality to Marketing Cloud integration

### LINKED ISSUE #2668: Add click to copy functionality to Marketing Cloud
On the Marketing Cloud "Content for email" admin page, I want to click one button from the dropdown to copy the URL of the desired content.

[Figma mockup for Quickstart](https://www.figma.com/file/KBr2TQ5YVGimwErjfMk4Xv/QS2-News-in-Marketing-Cloud?type=design&node-id=188-299&mode=design&t=McdIQRTpWJHZtxyo-0)
[Codepen for Marketing Cloud HTML](https://codepen.io/itsmemario/pen/GRXNPRL)

Follow up to https://github.com/az-digital/az_quickstart/issues/2644

### PR DESCRIPTION
**Pull request by danahertzberg**
## Description
On the Marketing Cloud "Content for email" admin page, I want to click one button from the dropdown to copy the URL of the desired content.

## Related issues
Closes #2668 

## How to test

1. Enable the `Quickstart for Marketing Cloud` module
2. Navigate to the `Content for email` page under Content

https://5cdfd524-ed42-4b88-839e-c17a43a91655.probo.build/

### Click to copy functionality

1. Click on a desired layout in the `Copy email snippet` dropdown
2. Ensure "Link copied!" pop up looks good
4. Ensure correct link was in fact copied to your clipboard (can paste somewhere)

### Related info content

1. Accordion on `Content for email` page is complete and descriptive
2. All layout options have an image and accurate description

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [x] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by danahertzberg** [comment]
Current status: Have admin table layout complete. Need help to figure out how to get click to copy functionality working. Some work is started but not successful so far.

**Comment by trackleft** [comment]
See https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Element%21Dropbutton.php/class/Dropbutton/9.3.x

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function buildOptionsForm(&$form, FormStateInterface $form_state) {
+    parent::buildOptionsForm($form, $form_state);
+
+    $form['dropbutton_type'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Dropbutton Type'),
+      '#description' => $this->t('Description of your custom option.'),
```
Comment (joeparsons) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:43]:
```suggestion
      '#description' => $this->t('Dropdown button type.'),
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      ],
+    ];
+    $form['click_action'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Click Action'),
+      '#description' => $this->t('Determine how links should behave.'),
+      '#default_value' => $this->options['click_action'],
+      '#options' => [
+        'none' => 'None',
+        'js-click2copy' => 'Click to Copy to Clipboard via JS',
```
Comment (joeparsons) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:57]: Nitpick:
```suggestion
        'js-click2copy' => 'Click to copy to clipboard via JS',
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      '#description' => $this->t('Description of your custom option.'),
+      '#default_value' => $this->options['dropbutton_type'],
+      '#options' => [
+        'small' => 'Small',
+        'extrasmall' => 'Extra Small',
+      ],
+    ];
+    $form['click_action'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Click Action'),
```
Comment (joeparsons) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:52]:
```suggestion
      '#title' => $this->t('Click action'),
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    parent::buildOptionsForm($form, $form_state);
+
+    $form['dropbutton_type'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Dropbutton Type'),
+      '#description' => $this->t('Description of your custom option.'),
+      '#default_value' => $this->options['dropbutton_type'],
+      '#options' => [
+        'small' => 'Small',
+        'extrasmall' => 'Extra Small',
```
Comment (joeparsons) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:47]:
```suggestion
        'extrasmall' => 'Extra small',
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildOptionsForm(&$form, FormStateInterface $form_state) {
+    parent::buildOptionsForm($form, $form_state);
+
+    $form['dropbutton_type'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Dropbutton Type'),
```
Comment (joeparsons) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:42]:
```suggestion
      '#title' => $this->t('Dropbutton type'),
```

File: modules/custom/az_marketing_cloud/config/install/views.view.az_marketing_cloud.yml [hunk 6]
```diff
@@ -1228,7 +1679,7 @@ display:
...
...ng_cloud/images/hero-layout.png\"></td>\r\n<td width=\"85%\"><p><strong>Hero Layout</strong><br>Your image and text take up the full width of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/50-50-layout.png\"></td>\r\n<td><p><strong>50/50 Layout</strong><br>\r\nYour image and text take up the same amount of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/30-70-layout.png\"></td>\r\n<td><p><strong>30/70 Layout</strong><br>\r\nYour image is square and takes up 30% of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/text-layout.png\"></td>\r\n<td><p><strong>Text Layout</strong><br>\r\nNo image is used. Your heading, summary and text link is used.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/event-layout.png\"></td>\r\n<td><p><strong>Event Layout</strong><br>\r\nNo image is used. Your date, heading and summary are used.</p></td>\r\n</tr>\r\n</table>\r\n</details>"
```
Comment (bberndt-uaz) [inline modules/custom/az_marketing_cloud/config/install/views.view.az_marketing_cloud.yml:1682]:
```suggestion
            value: "<p>Manage the content you would like to share with your Trellis Marketing Cloud email templates.</p>\r\n\r\n<details data-drupal-selector=\"edit-field-az-metatag-0\" class=\"accordion__item js-form-wrapper form-wrapper claro-details claro-details--accordion-item\" id=\"edit-field-az-metatag-0\" data-once=\"details\">\r\n\r\n<summary role=\"button\" aria-controls=\"edit-field-az-metatag-0\" aria-expanded=\"true\" aria-pressed=\"true\" class=\"claro-details__summary claro-details__summary--accordion-item\">How does this feature work?</summary>\r\n\r\n<div class=\"claro-details__wrapper details-wrapper claro-details__wrapper--accordion-item\">\r\n        <div class=\"claro-details__content claro-details__content--accordion-item\">\r\n    \r\n<p><em><strong>NOTE:</strong> This feature is intended to be used with Trellis Marketing Cloud.</em></p>\r\n\r\n<p>Use the <strong>Email Snippet</strong> link below to get the URL for the layout you would like to use in your email template. \LOnce copied, you can paste that URL into the External Content module on Marketing Cloud.</p>\r\n\r\n<p><strong>Layout options:</strong></p>\r\n<table>\r\n<tr>\r\n<td width=\"15%\"><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/hero-layout.png\"></td>\r\n<td width=\"85%\"><p><strong>Hero Layout</strong><br>Your image and text take up the full width of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/50-50-layout.png\"></td>\r\n<td><p><strong>50/50 Layout</strong><br>\r\nYour image and text take up the same amount of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/30-70-layout.png\"></td>\r\n<td><p><strong>30/70 Layout</strong><br>\r\nYour image is square and takes up 30% of horizontal space.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/text-layout.png\"></td>\r\n<td><p><strong>Text Layout</strong><br>\r\nNo image is used. Your heading, summary and text link are used.</p></td>\r\n</tr>\r\n<tr>\r\n<td><img src=\"/profiles/custom/az_quickstart/modules/custom/az_marketing_cloud/images/event-layout.png\"></td>\r\n<td><p><strong>Event Layout</strong><br>\r\nNo image is used. Your date, heading and summary are used.</p></td>\r\n</tr>\r\n</table>\r\n</details>"
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js [hunk 7]
```diff
@@ -1,22 +1,37 @@
...
-}
-
-copyLinks.forEach(element => (
-  element.addEventListener(
-    "click",
-    handleClick,
-    false
-  )
-));
+})(this, this.document);
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js:37]:
```suggestion
})(this, this.document);

```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    ];
+    if (!empty($dropbutton_type)) {
+      $dropbutton['#dropbutton_type'] = $dropbutton_type;
+    }
+    if (!empty($links)) {
+      $dropbutton['#links'] = $links;
+    }
+    if ($this->options['click_action'] == 'js-click2copy') {
+      $dropbutton['#attributes']['class'][] = $click_action;
+      $dropbutton['#attached']['library'][] = 'az_marketing_cloud/admin';
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:81]: What do you think of this?
```suggestion
      $dropbutton['#attached']['library'][] = 'az_core/dropbutton-click2copy';
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js [hunk 9]
```diff
@@ -1,22 +1,37 @@
...
+* See the following change record for more information,
+* https://www.drupal.org/node/2815083
+* @preserve
+**/
+(function (window, document) {
+  function init() {
+    var copyLinks = document.querySelectorAll('.js-click2copy a');
+    function _handleClick(event) {
+      var baseUrl = window.location.origin;
+      console.log(event);
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js:12]:
```suggestion
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    console.log(event);
+    if (event.type === 'click') {
+      event.preventDefault();
+      const href = event.target.getAttribute('href');
+      navigator.clipboard.writeText(baseUrl + href);
+      event.target.classList.add(
+        'js-click-copy--copied',
+        'action-link--icon-checkmark',
+      );
+      removeClass(event.srcElement)
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js:33]:
```suggestion
      removeClass(event.target)
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php (hunk 8 above)
Comment (danahertzberg) [reply modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:81]: @bberndt-uaz @joeparsons

**Comment by danahertzberg** [comment]
Here's what I got on the probo site. Click to copy works!! But got this huge notice after enabling the module.
image

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ */
+class AzDropbutton extends Links {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defineOptions() {
+    $options = parent::defineOptions();
+    $options['dropbutton_type'] = [
+      'small' => 'small',
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:24]:
```suggestion
      'default' => 'small',
```

File: modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+   * {@inheritdoc}
+   */
+  public function defineOptions() {
+    $options = parent::defineOptions();
+    $options['dropbutton_type'] = [
+      'default' => 'small',
+      'extrasmall' => 'extrasmall',
+    ];
+    $options['click_action'] = [
+      'none' => '',
```
Comment (trackleft) [inline modules/custom/az_marketing_cloud/src/Plugin/views/field/AzDropbutton.php:28]:
```suggestion
      'default' => '',
```

**Comment by trackleft** [comment]
FYI, to remove links that don't exist, and add click to copy functionality, you could probably extend this plugin https://git.drupalcode.org/project/drupal/-/blob/11.x/core/modules/views/src/Plugin/views/field/EntityLink.php

**Comment by danahertzberg** [comment]
As-is today, I'm unable to successfully install the Marketing Cloud module. I will require help on this one.

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+((window, document) => {
+  function init() {
+    const copyLinks = document.querySelectorAll('.js-click2copy a');
+
+    function removeClass(element) {
+      setTimeout(() => {
+        element.classList.remove(
+          'js-click-copy--copied',
+          'action-link--icon-checkmark',
```
Comment (danahertzberg) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js:14]:
```suggestion
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    function _handleClick(event) {
+      const baseUrl = window.location.origin;
+      if (event.type === 'click') {
+        event.preventDefault();
+        const href = event.target.getAttribute('href');
+        navigator.clipboard.writeText(baseUrl + href);
+        event.target.classList.add(
+          'js-click-copy--copied',
+          'action-link--icon-checkmark',
```
Comment (danahertzberg) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js:37]:
```suggestion
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js [hunk 15]
```diff
@@ -1,22 +1,36 @@
...
+* See the following change record for more information,
+* https://www.drupal.org/node/2815083
+* @preserve
+**/
+(function (window, document) {
+  function init() {
+    var copyLinks = document.querySelectorAll('.js-click2copy a');
+    function removeClass(element) {
+      setTimeout(function () {
+        element.classList.remove('js-click-copy--copied', 'action-link--icon-checkmark');
```
Comment (danahertzberg) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.js:12]:
```suggestion
        element.classList.remove('js-click-copy--copied');
```

**Comment by danahertzberg** [comment]
I think I have the CSS done! And the click-to-copy functionality is working on probo. I'd like to have that be the end of this PR to get core functionality merged in. Then have a separate PR for only showing pertinent links related to the content type in the dropdown.

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+((window, document) => {
+  function init() {
+    const copyLinks = document.querySelectorAll('.js-click2copy a');
+
+    function removeClass(element) {
+      setTimeout(() => {
+        element.classList.remove(
+          'js-click-copy--copied',
```
Comment (danahertzberg) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js:13]:
```suggestion
          'js-click-copy--copied'
```

File: modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+     * copied.
+     */
+    function _handleClick(event) {
+      const baseUrl = window.location.origin;
+      if (event.type === 'click') {
+        event.preventDefault();
+        const href = event.target.getAttribute('href');
+        navigator.clipboard.writeText(baseUrl + href);
+        event.target.classList.add(
+          'js-click-copy--copied',
```
Comment (danahertzberg) [inline modules/custom/az_marketing_cloud/js/az-marketing-cloud-admin.es6.js:35]:
```suggestion
          'js-click-copy--copied'
```

**Comment by danahertzberg** [comment]
And will need help with the linting errors. I have no idea what it's suggesting lol.

File: modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install [hunk 18]
```diff
@@ -5,8 +5,6 @@
  * Install, update and uninstall functions for az_paragraphs_text_media module.
  */
 
-use Drupal\paragraphs\Entity\Paragraph;
-
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install:9]: Joe added this back somewhere, so we should undo this before this PR is merged.

File: modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install [hunk 19]
```diff
@@ -5,8 +5,6 @@
  * Install, update and uninstall functions for az_paragraphs_text_media module.
  */
```
Comment (danahertzberg) [inline modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install:7]:
```suggestion

use Drupal\paragraphs\Entity\Paragraph;

```

**Comment by danahertzberg** [comment]
Ready for final review!

**Review by sanjanans — APPROVED** [review]
Everything looks good.

**Review by trackleft — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Pull request merged**